pkg/maven/maven.go

```diff
@@ -103,6 +127,130 @@ func Evaluate(pomFile, expression string, command mavenExecRunner) (string, erro
 	return value, nil
 }
 
+// InstallFile installs a maven artifact and its pom into the local maven repository.
+// If "file" is empty, only the pom is installed.
+// "pomFile" may not be empty.
+func InstallFile(file, pomFile string, command mavenExecRunner, utils mavenUtils, projectSettingsFile, globalSettingsFile, m2Path string) error {
+	if len(pomFile) == 0 {
+		return fmt.Errorf("pomFile can't be empty")
+	}
+
+	var defines []string
+	if len(file) > 0 {
+		defines = append(defines, "-Dfile="+file)
+	} else {
+		defines = append(defines, "-Dfile="+pomFile)
+	}
+	defines = append(defines, "-DpomFile="+pomFile)
+	mavenOptionsInstall := ExecuteOptions{
+		Flags:                       []string{},
+		Goals:                       []string{"install:install-file"},
+		Defines:                     defines,
+		PomPath:                     pomFile,
+		ProjectSettingsFile:         projectSettingsFile,
+		GlobalSettingsFile:          globalSettingsFile,
+		M2Path:                      m2Path,
+		LogSuccessfulMavenTransfers: false,
+	}
+	_, err := Execute(&mavenOptionsInstall, command)
+	return err
+}
+
+// InstallMavenArtifacts finds maven modules (identified by pom.xml files) and installs the artifacts into the local maven repository.
+func InstallMavenArtifacts(command mavenExecRunner, utils mavenUtils, p piperutils.FileUtils, m2Path, projectSettingsFile, globalSettingsFile string) error {
+	err := flattenPom(command, projectSettingsFile, globalSettingsFile, m2Path)
+	if err != nil {
+		return err
+	}
+
+	pomFiles, err := utils.glob(filepath.Join("**", "pom.xml"))
+	if err != nil {
+		return err
+	}
+
+	oldWorkingDirectory, err := utils.getwd()
+	if err != nil {
+		return err
+	}
+
+	for _, pomFile := range pomFiles {
+		log.Entry().Info("Installing maven artifacts from module: " + pomFile)
+		dir := utils.dir(pomFile)
+		err = utils.chdir(dir)
+		if err != nil {
+			return err
+		}
+
+		pomXmlBytes, err := p.FileRead("pom.xml")
+		if err != nil {
+			return err
+		}
+		pomXmlString := string(pomXmlBytes)
+		if strings.Contains(pomXmlString, "<packaging>pom</packaging>") {
```

Comment:
Why not evaluate the packaging via Evaluate()?

Reply:
I tried to get away with the lazy way, but I see that this won't work in all cases. Probably still good enough?

Reply:
You replace these lines:

```golang
pomXmlBytes, err := p.FileRead("pom.xml")
if err != nil {
    return err
}
pomXmlString := string(pomXmlBytes)
if strings.Contains(pomXmlString, "<packaging>pom</packaging>") {
```

with these lines:

```golang
packaging, err := Evaluate("pom.xml", "project.packaging")
if err != nil {
    return err
}
if packaging == "pom") {
```

You even save one line of code, and it is more robust. I don't see how this is not better.